### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by yodaiken**

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    }]
+  ],
+  "plugins": [
+    "transform-async-generator-functions",
+    "transform-flow-strip-types",
+    ["transform-object-rest-spread", { "useBuiltIns": true }],
+    "transform-class-properties",
+    "add-module-exports"
+  ]
+}
```
Comment (jquense) [inline .babelrc]: didn't like my scaffolding ? :P

Comment (jquense) [reply .babelrc]: (but seriously can we use @4c/babel-preset-4catalyzer ?)

File: src/SubscriptionServer.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      wsEngine: 'ws',
+    });
+
+    this.io.on('connection', socket => {
+      const request = Object.create((express: any).request);
+      Object.assign(request, socket.request);
+
+      const context = config.createContext(request);
+
+      // eslint-disable-next-line no-new
```
Comment (jquense) [inline src/SubscriptionServer.js]: what is this rule?

File: src/TenantSocketConnection.js [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    this.renewTimer = setTimeout(() => {
+      this.handleAuthenticate(implicitToken);
+    }, deltaMSAdjusted);
+  }
+
+  async updateCredentials(implicitToken: string) {
+    this.credentials = await this.config.fetchCredentials(
+      this.config.context,
+      implicitToken,
+    );
```
Comment (jquense) [inline src/TenantSocketConnection.js]: @taion is your thought we should eventually split most of this stuff out into a CredentialManager?

File: package.json [hunk 4]
```diff
@@ -0,0 +1,40 @@
+{
+  "name": "graphql-subscription-server",
```
Comment (jquense) [inline package.json]: should we scope this? `@4c/graphql-subscription-server`

Comment (taion) [reply package.json]: :+1:

File: src/TenantSocketConnection.js (hunk 3 above)
Comment (taion) [reply src/TenantSocketConnection.js]: yeah, it should look like:

```js
type CredentialsManager<Credentials> {
  updateAuthorization: (authorization: string) => void,
  getCredentials: () => Credentials,
  close: () => void, // Or maybe some async thing.
};
```

And you pass in `context` when you construct it via a:

```js
createCredentialsManager: (context: Context) => CredentialsManager<Credentials>,
```

File: .babelrc (hunk 1 above)
Comment (yodaiken) [reply .babelrc]: wasn't ready yet :P

File: src/SubscriptionServer.js (hunk 2 above)
Comment (yodaiken) [reply src/SubscriptionServer.js]: using `new` and not storing the result. think it makes sense to do here, though.

File: .babelrc (hunk 1 above)
Comment (jquense) [reply .babelrc]: my text got swallowed. no scaffolding is fine , but can we use the 4c babel-preset here

**Comment by jquense** [comment]
NINJA COMMITS

**Comment by taion** [comment]
@jquense Did you see https://github.com/ButterflyNetwork/olympus-graphql/pull/288?

**Comment by taion** [comment]
^ that has the credentialsmanager stuff

File: LICENSE [hunk 5]
```diff
@@ -0,0 +1,22 @@
+The MIT License (MIT)
+
+Copyright (c) 2016 4Catalyzer
```
Comment (taion) [inline LICENSE]: did we really write this in 2016?

File: src/TenantSocketConnection.js [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+  if (cb) cb();
+};
+
+function isExpired(expSeconds: ?number): boolean {
+  return expSeconds != null && !!(Date.now() > expSeconds * SECONDS_TO_MS);
+}
+
+// TenantSocketConnection manages a single socket connection for a single user.
+// Includes
+//  - authorization,
```
Comment (taion) [inline src/TenantSocketConnection.js]: nb fix this comment when we do the `CredentialsManager` thing

File: src/TenantSocketConnection.js [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+    this.credentials.tenants && !isExpired(this.credentials.expSeconds);
+
+  isAuthorized = (data: any) =>
+    !!(
+      this.isAuthenticated() &&
+      this.config.hasPermission(data, this.credentials)
+    );
+
+  scheduleRenew(implicitToken: string) {
+    const deltaMS = this.credentials.expSeconds * SECONDS_TO_MS - Date.now();
```
Comment (taion) [inline src/TenantSocketConnection.js]: `deltaMs`?

**Pull request merged**